### PR TITLE
[SPARK-30200][SQL] Add ExplainMode for Dataset.explain

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

This pr intends to add `ExplainMode` for explaining `Dataset/DataFrame` with a given format mode (`ExplainMode`). `ExplainMode` has four types along with the SQL EXPLAIN command: `Simple`, `Extended`, `Codegen`, `Cost`, and `Formatted`.

For example, this pr enables users to explain DataFrame/Dataset with the `FORMATTED` format implemented in #24759;
```
scala> spark.range(10).groupBy("id").count().explain(ExplainMode.Formatted)
== Physical Plan ==
* HashAggregate (3)
+- * HashAggregate (2)
   +- * Range (1)

(1) Range [codegen id : 1]
Output: [id#0L]
     
(2) HashAggregate [codegen id : 1]
Input: [id#0L]
     
(3) HashAggregate [codegen id : 1]
Input: [id#0L, count#8L]
```

This comes from [the @cloud-fan suggestion.](https://github.com/apache/spark/pull/24759#issuecomment-560211270)

### Why are the changes needed?

To follow the SQL EXPLAIN command.

### Does this PR introduce any user-facing change?

No, this is just for a new API in Dataset.

### How was this patch tested?

Add tests in `ExplainSuite`.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -522,36 +522,58 @@ class Dataset[T] private[sql](
...
   /**
-   * Prints the plans (logical and physical) to the console for debugging purposes.
+   * Prints the plans (logical and physical) with a format specified by a given explain mode.
    *
    * @group basic
-   * @since 1.6.0
+   * @since 3.0.0
    */
-  def explain(extended: Boolean): Unit = {
+  def explain(mode: ExplainMode): Unit = {
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: How about retain the old api and add a new api ?

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: @ulysses-you . @maropu already did. Please see [line 564](https://github.com/apache/spark/pull/26829/files#diff-7a46f10c3cedbf013cf255564d9483cdR564).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 2]
```diff
@@ -132,13 +131,15 @@ case class DataWritingCommandExec(cmd: DataWritingCommand, child: SparkPlan)
  * (but do NOT actually execute it).
  *
  * {{{
- *   EXPLAIN (EXTENDED | CODEGEN) SELECT * FROM ...
+ *   EXPLAIN (EXTENDED | CODEGEN | COST | FORMATTED) SELECT * FROM ...
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: Thank you for fixing this together.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Yea, we should keep this. Thanks for the comment, @dongjoon-hyun

**Comment by maropu** [comment]
Thanks for the check, @dongjoon-hyun !

File: sql/core/src/main/java/org/apache/spark/sql/ExplainMode.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  Simple,
+  /**
+   * Extended mode means that when printing explain for a DataFrame, both logical and physical
+   * plans are expected to be printed to the console.
+   *
+   * @since 3.0.0
+   */
+  Extended,
+  /**
+   * Extended mode means that when printing explain for a DataFrame, if generated codes are
```
Comment (ulysses-you) [inline sql/core/src/main/java/org/apache/spark/sql/ExplainMode.java]: Codegen mode ?

File: sql/core/src/main/java/org/apache/spark/sql/ExplainMode.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+   * available, a physical plan and the generated codes are expected to be printed to the console.
+   *
+   * @since 3.0.0
+   */
+  Codegen,
+  /**
+   * Extended mode means that when printing explain for a DataFrame, if plan node statistics are
```
Comment (ulysses-you) [inline sql/core/src/main/java/org/apache/spark/sql/ExplainMode.java]: The same.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Oh, I see it.

File: sql/core/src/main/java/org/apache/spark/sql/ExplainMode.java (hunk 3 above)
Comment (maropu) [reply sql/core/src/main/java/org/apache/spark/sql/ExplainMode.java]: oh... I'll do follow-up, thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 5]
```diff
@@ -522,36 +522,58 @@ class Dataset[T] private[sql](
...
     // scalastyle:off println
     println(outputString)
     // scalastyle:on println
   }
 
+  /**
+   * Prints the plans (logical and physical) to the console for debugging purposes.
+   *
+   * @group basic
+   * @since 1.6.0
```
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:562]: That should really be 3.0.0, no?

**Pull request closed without merging**